features/steps/steps.py

```diff
@@ -200,6 +209,61 @@ def step_impl(context, constraint, num, entity):
 
     handle_errors(context, errors)
 
+@then('Each {entity} must nest {constraint} {num:d} instance(s) of {other_entity}')
+def step_impl(context, entity, num, constraint, other_entity):
+    stmt_to_op = {'exactly': operator.eq, "at most": operator.le}
+    assert constraint in stmt_to_op
+    op = stmt_to_op[constraint]
+
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [entity for rel in inst.IsNestedBy for entity in rel.RelatedObjects]
+            if not op(sum(1 for i in nested_entities if i.is_a() == other_entity), num):
+                errors.append(instance_count_error([i for i in nested_entities if i.is_a() == other_entity]))
+
+    handle_errors(context, errors)
+
+@then('Each {entity} must be nested only by {other_entity}')
+def step_impl(context, entity, other_entity):
+    
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            relating_objects_lst = [i.RelatingObject.is_a() for i in inst.Nests]
+            relating_object = relating_objects_lst[0]
+            if not other_entity == relating_object:
+                errors.append(instance_nesting_error(inst, relating_objects_lst))
+
+@then('Each {entity} must nest only the following entities: {other_entities}')
+def step_impl(context, entity, other_entities):
+
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [i for rel in inst.IsNestedBy for i in rel.RelatedObjects]
+            if not set([i.is_a() for i in nested_entities]).issubset(other_entities.split(',')):
+                errors.append(instance_nesting_error(inst, nested_entities))
+                #todo differentiate which 'prohibited' entity was found
+
+@then('Each {entity} nests a list of {segment_annotation}, each of which has DesignParameters typed as {parameter_annotation}')
+def step_impl(context, entity, segment_annotation, parameter_annotation):
+
+    errors = []
+    
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            segments = [inst for rel in inst.IsNestedBy for inst in rel.RelatedObjects]
+            if not all(operator.eq(segment.is_a(),segment_annotation) for segment in segments):
+                errors.append(instance_nesting_error(inst, segments))
+            parameters = [segment.DesignParameters for segment in segments]
+            if not all(operator.eq(parameters.is_a(),parameter_annotation) for parameters in parameters):
+                errors.append(instance_nesting_error(inst, parameters))
+            #@todo perhaps a different data class with more precise information, how many? which other attributes were found?
```

Comment:
Indeed, that's exactly the reason why I find my approach to the gherkin clearer. The error category would be much more natural.